File: backend/src/leetcode/leetcode.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { LeetcodeEntry, LeetcodeDifficulty, LeetcodeStatus } from './leetcode-entry.entity';
import { CreateEntryDto } from './dto/create-entry.dto';
import { UpdateEntryDto } from './dto/update-entry.dto';

@Injectable()
export class LeetcodeService {
  constructor(
    @InjectRepository(LeetcodeEntry)
    private readonly repo: Repository<LeetcodeEntry>,
  ) {}

  create(dto: CreateEntryDto): Promise<LeetcodeEntry> {
    const entry = this.repo.create({
      ...dto,
      url: dto.url ?? null,
      status: dto.status ?? LeetcodeStatus.ATTEMPTED,
      topics: dto.topics ?? [],
      notes: dto.notes ?? null,
      solvedAt:
        dto.status === LeetcodeStatus.SOLVED
          ? new Date().toISOString().slice(0, 10)
          : null,
    });
    return this.repo.save(entry);
  }

  async findAll(params: { status?: LeetcodeStatus; difficulty?: LeetcodeDifficulty; topic?: string }) {
    const all = await this.repo.find({ order: { updatedAt: 'DESC' } });
    return all.filter((e) => {
      if (params.status && e.status !== params.status) return false;
      if (params.difficulty && e.difficulty !== params.difficulty) return false;
      if (
        params.topic &&
        !e.topics.map((t) => t.toLowerCase()).includes(params.topic.toLowerCase())
      ) {
        return false;
      }
      return true;
    });
  }

  async findOne(id: string): Promise<LeetcodeEntry> {
    const entry = await this.repo.findOne({ where: { id } });
    if (!entry) throw new NotFoundException(`LeetCode entry ${id} not found`);
    return entry;
  }

  async update(id: string, dto: UpdateEntryDto): Promise<LeetcodeEntry> {
    const entry = await this.findOne(id);
    const wasSolved = entry.status === LeetcodeStatus.SOLVED;
    Object.assign(entry, dto);
    if (!wasSolved && entry.status === LeetcodeStatus.SOLVED && !entry.solvedAt) {
      entry.solvedAt = new Date().toISOString().slice(0, 10);
    }
    return this.repo.save(entry);
  }

  async remove(id: string): Promise<void> {
    const entry = await this.findOne(id);
    await this.repo.remove(entry);
  }

  async stats() {
    const all = await this.repo.find();
    const solved = all.filter((e) => e.status === LeetcodeStatus.SOLVED);

    const byDifficulty = (difficulty: LeetcodeDifficulty) => ({
      solved: solved.filter((e) => e.difficulty === difficulty).length,
      attempted: all.filter((e) => e.difficulty === difficulty).length,
    });

    const topicCounts: Record<string, number> = {};
    for (const e of solved) {
      for (const t of e.topics) {
        topicCounts[t] = (topicCounts[t] || 0) + 1;
      }
    }

    return {
      totalProblems: all.length,
      totalSolved: solved.length,
      easy: byDifficulty(LeetcodeDifficulty.EASY),
      medium: byDifficulty(LeetcodeDifficulty.MEDIUM),
      hard: byDifficulty(LeetcodeDifficulty.HARD),
      topTopics: Object.entries(topicCounts)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .map(([topic, count]) => ({ topic, count })),
    };
  }
}

File: backend/src/main/java/com/devagent/leetcode/LeetcodeService.java
package com.devagent.leetcode;

import com.devagent.leetcode.LeetcodeEntry.Difficulty;
import com.devagent.leetcode.LeetcodeEntry.Status;
import com.devagent.leetcode.dto.CreateLeetcodeRequest;
import com.devagent.leetcode.dto.UpdateLeetcodeRequest;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class LeetcodeService {

    private final LeetcodeRepository repo;

    public LeetcodeService(LeetcodeRepository repo) {
        this.repo = repo;
    }

    public LeetcodeEntry create(String userId, CreateLeetcodeRequest req) {
        boolean isSolved = Status.Solved.equals(req.getStatus());
        LeetcodeEntry e = new LeetcodeEntry();
        e.setUserId(userId);
        e.setTitle(req.getTitle());
        e.setUrl(req.getUrl());
        e.setDifficulty(req.getDifficulty());
        e.setStatus(req.getStatus() != null ? req.getStatus() : Status.Attempted);
        e.setTopics(req.getTopics() != null ? req.getTopics() : List.of());
        e.setNotes(req.getNotes());
        e.setSolvedAt(isSolved ? LocalDate.now().toString() : null);
        return repo.save(e);
    }

    public List<LeetcodeEntry> findAll(String userId, String statusStr,
                                       String difficultyStr, String topic) {
        List<LeetcodeEntry> all = repo.findByUserIdOrderByUpdatedAtDesc(userId);
        return all.stream()
                .filter(e -> statusStr == null || e.getStatus().name().equalsIgnoreCase(statusStr))
                .filter(e -> difficultyStr == null || e.getDifficulty().name().equalsIgnoreCase(difficultyStr))
                .filter(e -> topic == null || e.getTopics().stream()
                        .anyMatch(t -> t.equalsIgnoreCase(topic)))
                .collect(Collectors.toList());
    }

    public List<LeetcodeEntry> findAllAdmin() {
        return repo.findAllByOrderByUpdatedAtDesc();
    }

    public LeetcodeEntry findOne(String userId, String id) {
        return repo.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "LeetCode entry " + id + " not found"));
    }

    public LeetcodeEntry update(String userId, String id, UpdateLeetcodeRequest req) {
        LeetcodeEntry e = findOne(userId, id);
        boolean wasSolved = Status.Solved.equals(e.getStatus());

        if (req.getTitle() != null) e.setTitle(req.getTitle());
        if (req.getUrl() != null) e.setUrl(req.getUrl());
        if (req.getDifficulty() != null) e.setDifficulty(req.getDifficulty());
        if (req.getStatus() != null) e.setStatus(req.getStatus());
        if (req.getTopics() != null) e.setTopics(req.getTopics());
        if (req.getNotes() != null) e.setNotes(req.getNotes());

        if (!wasSolved && Status.Solved.equals(e.getStatus()) && e.getSolvedAt() == null) {
            e.setSolvedAt(LocalDate.now().toString());
        }
        return repo.save(e);
    }

    public void remove(String userId, String id) {
        LeetcodeEntry e = findOne(userId, id);
        repo.delete(e);
    }

    public void adminRemove(String id) {
        repo.findById(id).ifPresentOrElse(repo::delete,
                () -> { throw new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "LeetCode entry " + id + " not found"); });
    }

    public Map<String, Object> stats(String userId) {
        List<LeetcodeEntry> all = repo.findByUserId(userId);
        List<LeetcodeEntry> solved = all.stream()
                .filter(e -> Status.Solved.equals(e.getStatus())).toList();

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("totalProblems", all.size());
        result.put("totalSolved", solved.size());
        result.put("easy", diffStats(all, solved, Difficulty.Easy));
        result.put("medium", diffStats(all, solved, Difficulty.Medium));
        result.put("hard", diffStats(all, solved, Difficulty.Hard));

        // Top topics
        Map<String, Long> topicCounts = solved.stream()
                .flatMap(e -> e.getTopics().stream())
                .collect(Collectors.groupingBy(t -> t, Collectors.counting()));

        List<Map<String, Object>> topTopics = topicCounts.entrySet().stream()
                .sorted(Map.Entry.<String, Long>comparingByValue().reversed())
                .limit(10)
                .map(entry -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("topic", entry.getKey());
                    m.put("count", entry.getValue());
                    return m;
                })
                .toList();
        result.put("topTopics", topTopics);
        return result;
    }

    private Map<String, Integer> diffStats(List<LeetcodeEntry> all,
                                            List<LeetcodeEntry> solved,
                                            Difficulty d) {
        return Map.of(
                "solved", (int) solved.stream().filter(e -> d.equals(e.getDifficulty())).count(),
                "attempted", (int) all.stream().filter(e -> d.equals(e.getDifficulty())).count()
        );
    }
}
